[FEAT] CLI overhaul, stdin support, some refactors

sbresin opened the pull request:
## description
originally i just wanted to add stdin support to integrate afmt into neovim through efm-langserver, but then it kind of escalated quickly. if wanted i could also only take parts of the changes into seperate PRs.

in order to not spend more time going in a possibly unwanted direction with this, i am putting this up for early feedback. 

The new CLI is very similar to prettier's and opens some interesting possibilities, that i want to add to the README as well if you want this change in the first place, like:
- using `--check` and a bash 3 liner to annotate unformatted code in a github action
- configuring afmt in efm-langserver to use in neovim or other editors

Looking forward to your feedback, thanks for contributing this tool and taking it open-source!

sbresin committed:
wip: modes, uni cli

sbresin committed:
feat: add stdin cli option + streamline args

sbresin committed:
feat: complex format return type, simpler threading code, stdin support

sbresin committed:
feat: terminal output for machines and humans

sbresin committed:
chore: adapt existing changes to new fn signatures

sbresin committed:
chore: update readme with new usage and examples

xixiaofinland commented:
Hi Sebastian,

Thank you for creating the PR! This being the first one, it’s a significant milestone and truly means a lot to me.

You’re definitely steering this in the right direction. Supporting stdin input (e.g., afmt < ./file.cls) is generally a best practice for CLI tools, and I agree we should implement it.

I noticed the "escalation" you mentioned. There are a few prerequisite code changes, including updates to the function signatures in the tests, which you’ve already started addressing. I also quickly went through some of your commits, they seem to be quite well-considered.

In addition, I agree that splitting the changes into multiple PRs is a good approach. It will make it easier to review and understand the scope of each change.

I have a 3-week holiday coming soon without a computer. I hope it's OK that I come back for it after. 

Looking forward to the next steps!

xixiaofinland commented:
Also FYI,  

1. I removed the "PR branch name" in the first failing GitHub action. 
2. Not sure why "Battle Tests (pull_request)" fails, it works for my own PR. Perhaps it's the action script issue. I will look into it in the future.

You may rebase the latest main branch to get the up-to-date actions.

sbresin commented:
thanks for your kind response :) cool that you are on board with the ideas. 

> Speaking of "efm-langserver" support, could you elaborate on what features you'd like to have? If they are general purpose features to better integrate into Nvim/VsCode/Other editors, I think we should scope them in.

[efm-langserver](https://github.com/mattn/efm-langserver) is a very useful tool, that allows to put any formatting/linting tool behind a lsp process/interface, so any editor that integrates with lsps can seamlessly integrate with it. The moment afmt supports stdin/out, it can already be used with efm-langserver and i already tested it in neovim, heres a shorter excerpt:
```lua
local afmt = {
    formatCommand = "afmt",
    formatStdin = true,
}

return {
    {
        "neovim/nvim-lspconfig",
        ---@class PluginLspOpts
        opts = {
            servers = {
                efm = {
                    init_options = {
                        documentFormatting = true,
                    },
                    filetypes = {
                        "apexcode", 
                    },
                    settings = {
                        rootMarkers = {".git/"},
                        lintDebounce = "3s",
                        languages = {
                            apexcode = {
                                afmt,
                            },
                        }
                    }
                }
            }
        }
    }
}
```
Theres also a VSCode Plugin for efm, that i didn't try though. In neovim it works really great.

> In addition, I agree that splitting the changes into multiple PRs is a good approach. It will make it easier to review and understand the scope of each change.
>
> I have a 3-week holiday coming soon without a computer. I hope it's OK that I come back for it after.

thats good! i'll have little time to work on this anyways and during the 3 weeks, i can tidy this up and split it into multiple incremental PRs. :) 

> I removed the "PR branch name" in the first failing GitHub action so the new version should pass now.

cool thank you, i could have renamed the branch as well, but i guess enforcing conventional commits is good enough and branch names don't matter this much after merging. :+1: 


I'll rebase and check the battle-test thing sometime this week, it fails because i changed the cli args and the output as well, but should be trivial to fix.

xixiaofinland commented:
We should definitely support seamless integration with tools like efm-langserver.

I’ve been considering how to draw a clear line between the functionalities of afmt and an editor plugin.

We already have an [open issue](https://github.com/xixiaofinland/afmt/issues/37) discussing Neovim support, where I shared a basic Lua configuration for using afmt.

Additionally, I’ve expressed interest in developing a VSCode plugin in [this issue](https://github.com/xixiaofinland/afmt/issues/83), especially since VSCode is almost used by all Salesforce developers.

sbresin committed:
fix: adapt battle test script to use stdin mode

sbresin commented:
> Great to hear that stdin/out can go far already!
I was only considering how to draw a clear line between the functionalities of afmt and separate editor plugins for afmt.

:+1: the great thing about efm is, that no dedicated plugin is needed for afmt support. In neovim it all works through the built-in lsp support. Also the current buffer can be formatted without writing or touching the file.
For VSCode users though it probably makes sense to have a dedicated plugin, that does not require additional configuration. i've built vscode extensions before and this one shouldn't be too hard, but having stdin/stdout support first would be helpful there as well. 

i have rebased and adapted the battle test script, so the checks should work now :crossed_fingers: 

Next slot of free time i have for this, i'll start open the first smaller incremental PRs.

xixiaofinland commented:
Fantastic! I will take a closer look at efm when back. It looks like a cool project. 

And I will come back to this thread after I’m back to the computer. Thanks a lot and Cheers!

xixiaofinland commented:
@sbresin  Hi Seba, I'm back. How shall we move on on this activity?
